File: ios/SsdidWallet/Domain/Notify/LocalNotificationStorage.swift
import Foundation

/// Persists in-app notifications to a JSON file via FileManager.
/// Published properties drive SwiftUI reactivity.
@MainActor
final class LocalNotificationStorage: ObservableObject {

    @Published private(set) var notifications: [LocalNotification] = []

    var unreadCount: Int {
        notifications.filter { !$0.isRead }.count
    }

    /// Maximum stored notifications. Oldest are evicted on save.
    static let maxNotifications = 500

    private let fileURL: URL

    init() {
        let docs = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        self.fileURL = docs.appendingPathComponent("local_notifications.json")
        self.notifications = Self.load(from: fileURL)
    }

    func save(_ notification: LocalNotification) {
        // Preserve isRead if notification already exists locally (avoids resurrecting read notifications on re-fetch)
        let existing = notifications.first { $0.id == notification.id }
        let merged = existing != nil ? LocalNotification(
            id: notification.id,
            mailboxId: notification.mailboxId,
            identityName: notification.identityName,
            payload: notification.payload,
            priority: notification.priority,
            receivedAt: notification.receivedAt,
            isRead: existing!.isRead
        ) : notification
        notifications.removeAll { $0.id == notification.id }
        notifications.insert(merged, at: 0)
        // Trim to capacity
        if notifications.count > Self.maxNotifications {
            notifications = Array(notifications.prefix(Self.maxNotifications))
        }
        persist()
    }

    func markAsRead(_ id: String) {
        guard let index = notifications.firstIndex(where: { $0.id == id }) else { return }
        notifications[index].isRead = true
        persist()
    }

    func markAllAsRead() {
        for i in notifications.indices {
            notifications[i].isRead = true
        }
        persist()
    }

    func delete(_ id: String) {
        notifications.removeAll { $0.id == id }
        persist()
    }

    // MARK: - Private

    private func persist() {
        let snapshot = notifications
        let url = fileURL
        Task.detached(priority: .utility) {
            do {
                let data = try JSONEncoder().encode(snapshot)
                try data.write(to: url, options: .atomic)
            } catch {
                #if DEBUG
                print("LocalNotificationStorage: failed to persist — \(error)")
                #endif
            }
        }
    }

    private static func load(from url: URL) -> [LocalNotification] {
        guard let data = try? Data(contentsOf: url) else { return [] }
        return (try? JSONDecoder().decode([LocalNotification].self, from: data)) ?? []
    }
}

File: android/sdk/ssdid-core/src/main/java/my/ssdid/sdk/platform/notify/LocalNotificationStorage.kt
package my.ssdid.sdk.platform.notify

import android.content.Context
import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import my.ssdid.sdk.domain.notify.LocalNotification
import my.ssdid.sdk.domain.notify.LocalNotificationStore

internal val Context.localNotificationsStore: DataStore<Preferences> by preferencesDataStore(name = "ssdid_local_notifications")

class LocalNotificationStorage(
    private val context: Context
) : LocalNotificationStore {
    private val key = stringPreferencesKey("notifications_json")
    private val json = Json { ignoreUnknownKeys = true }

    companion object {
        /** Maximum stored notifications. Oldest are evicted on save. */
        const val MAX_NOTIFICATIONS = 500
    }

    val allNotifications: Flow<List<LocalNotification>> =
        context.localNotificationsStore.data.map { prefs ->
            val raw = prefs[key] ?: return@map emptyList()
            runCatching { json.decodeFromString<List<LocalNotification>>(raw) }.getOrElse { emptyList() }
        }

    val unreadCount: Flow<Int> =
        allNotifications.map { list -> list.count { !it.isRead } }

    override suspend fun save(notification: LocalNotification) {
        context.localNotificationsStore.edit { prefs ->
            val current = prefs[key]?.let { runCatching { json.decodeFromString<List<LocalNotification>>(it) }.getOrElse { emptyList() } } ?: emptyList()
            // Preserve isRead if notification already exists locally (avoids resurrecting read notifications on re-fetch)
            val existing = current.firstOrNull { it.id == notification.id }
            val merged = if (existing != null) notification.copy(isRead = existing.isRead) else notification
            val updated = (current.filter { it.id != notification.id } + merged).takeLast(MAX_NOTIFICATIONS)
            prefs[key] = json.encodeToString(updated)
        }
    }

    override suspend fun markAsRead(id: String) {
        context.localNotificationsStore.edit { prefs ->
            val current = prefs[key]?.let { runCatching { json.decodeFromString<List<LocalNotification>>(it) }.getOrElse { emptyList() } } ?: return@edit
            val updated = current.map { if (it.id == id) it.copy(isRead = true) else it }
            prefs[key] = json.encodeToString(updated)
        }
    }

    override suspend fun markAllAsRead() {
        context.localNotificationsStore.edit { prefs ->
            val current = prefs[key]?.let { runCatching { json.decodeFromString<List<LocalNotification>>(it) }.getOrElse { emptyList() } } ?: return@edit
            val updated = current.map { it.copy(isRead = true) }
            prefs[key] = json.encodeToString(updated)
        }
    }

    override suspend fun delete(id: String) {
        context.localNotificationsStore.edit { prefs ->
            val current = prefs[key]?.let { runCatching { json.decodeFromString<List<LocalNotification>>(it) }.getOrElse { emptyList() } } ?: return@edit
            val updated = current.filter { it.id != id }
            prefs[key] = json.encodeToString(updated)
        }
    }
}
